fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -513,17 +518,63 @@ private ScoredPlan planFilter(@Nonnull PlanContext planContext, @Nonnull QueryCo
         return withInJoin;
     }
 
+    @Nullable
     private ScoredPlan planFilterWithInJoin(@Nonnull PlanContext planContext, @Nonnull InExtractor inExtractor, boolean needOrdering) {
-        final ScoredPlan bestPlan = planFilterForInJoin(planContext, inExtractor.subFilter(), needOrdering);
-        if (bestPlan != null) {
-            final RecordQueryPlan wrapped = inExtractor.wrap(planContext.rankComparisons.wrap(bestPlan.plan, bestPlan.includedRankComparisons, metaData));
-            final ScoredPlan scoredPlan = new ScoredPlan(bestPlan.score, wrapped);
-            if (needOrdering) {
-                scoredPlan.planOrderingKey = inExtractor.adjustOrdering(bestPlan.planOrderingKey, false);
+        int maxNumReplans = getConfiguration().getMaxNumReplansForInToJoin();
+        int numReplan = 0;
+        boolean progress = true;
+        ScoredPlan bestPlan = null;
+        while (numReplan <= maxNumReplans) {
+            bestPlan = planFilterForInJoin(planContext, inExtractor.subFilter(), needOrdering);
+            if (bestPlan != null) {
+                final Set<String> inBindings = inExtractor.getInBindings();
+                final Set<String> sargedInBindings = bestPlan.getSargedInBindings();
+                if (sargedInBindings.containsAll(inBindings)) {
+                    break;
+                }
+
+                // create a new in extractor that only uses the in-clauses we were actually able to use
+                inExtractor = inExtractor.filter((componentWithComparison, inBinding) -> {
+                    if (sargedInBindings.contains(inBinding)) {
+                        return true;
+                    }
+                    // we cannot filter out INs that are defined using more complicated things like "rank(...) IN (...)"
+                    return componentWithComparison instanceof QueryRecordFunctionWithComparison;
+                });
+
+                if (inBindings.size() == inExtractor.getInBindings().size()) {
+                    // we were unable to make progress
+                    progress = false;
+                    break;
+                }
+                
+                // Continue to re-plan with fewer in-clauses or exit the loop.
+                numReplan ++;
+            } else {
+                return null;
```

Comment:
Since this is an early exit, I suggest inverting the `if` and getting back some indentation.

It looks like this covers the case we discussed with IN-Union plans due to forced ordering, since that is stable `null` no matter how the filters are changed.

Reply:
done